internal/vault/client_factory.go

```diff
@@ -572,12 +661,76 @@ func (m *cachingClientFactory) restoreAllRequest(ctx context.Context, client ctr
 	return req, nil
 }
 
-func (c *cachingClientFactory) incrementRequestCounter(operation string, err error) {
+func (m *cachingClientFactory) incrementRequestCounter(operation string, err error) {
 	if err != nil {
-		c.requestErrorCounterVec.WithLabelValues(operation).Inc()
+		m.requestErrorCounterVec.WithLabelValues(operation).Inc()
 	} else {
-		c.requestCounterVec.WithLabelValues(operation).Inc()
+		m.requestCounterVec.WithLabelValues(operation).Inc()
+	}
+}
+
+func (m *cachingClientFactory) startClientCallbackHandler(ctx context.Context) {
+	logger := m.logger.WithName("clientCallbackHandler")
+	if m.callbackHandlerCancel != nil {
+		logger.Info("Already started")
+		return
 	}
+
+	callbackCtx, cancel := context.WithCancel(ctx)
+	m.callbackHandlerCancel = cancel
+
+	logger.Info("Starting client callback handler")
+
+	go func() {
+		if m.callbackHandlerCh == nil {
+			m.callbackHandlerCh = make(chan Client)
+		}
+		defer func() {
+			m.callbackHandlerCh = nil
+			close(m.callbackHandlerCh)
+		}()
+
+		for {
+			select {
+			case <-callbackCtx.Done():
+				logger.Info("Client callback handler done")
+				return
+			case c := <-m.callbackHandlerCh:
+				if c.IsClone() {
+					continue
+				}
+
+				cacheKey, err := c.GetCacheKey()
+				if err != nil {
+					logger.Error(err, "Invalid client, client callbacks not executed",
+						"cacheKey", cacheKey)
+					continue
+				}
+
+				// remove the client from the cache, it will be recreated when a reconciler
+				// requests it.
+				logger.V(consts.LogLevelDebug).Info("Removing client from cache", "cacheKey", cacheKey)
+				m.cache.Remove(cacheKey)
```

Comment:
@tvoran - instead of logging the client back in, the better approach was to remove it from the cache and then let the next reconciler retrieve it if need be. That ensure that any clones are properly purged in the cross namespace scenario and the new client is properly configured. The cache key may change between removal and the next reconcile, so the handled client being evicted from the cache serves two purposes.